nexus/get_info_test.go

```diff
@@ -64,3 +65,123 @@ func TestGetInfoHandleFromClientNoHeader(t *testing.T) {
 	require.Equal(t, handle.ID, info.ID)
 	require.Equal(t, OperationStateCanceled, info.State)
 }
+
+type asyncWithInfoTimeoutHandler struct {
+	UnimplementedHandler
+}
+
+func (h *asyncWithInfoTimeoutHandler) StartOperation(ctx context.Context, operation string, input *LazyValue, options StartOperationOptions) (HandlerStartOperationResult[any], error) {
+	return &HandlerStartOperationResultAsync{
+		OperationID: "needs /URL/ escaping",
+	}, nil
+}
+
+func (h *asyncWithInfoTimeoutHandler) GetOperationInfo(ctx context.Context, operation, operationID string, options GetOperationInfoOptions) (*OperationInfo, error) {
+	time.Sleep(20 * time.Millisecond)
+
+	if ctx.Err() != nil {
+		return nil, HandlerErrorf(HandlerErrorTypeDownstreamTimeout, "handler exceeded request timeout of %s", options.Header.Get(headerRequestTimeout))
+	}
+
+	return &OperationInfo{
+		ID:    operationID,
+		State: OperationStateCanceled,
+	}, nil
+}
+
+func TestGetInfoHandlerTimeout(t *testing.T) {
+	ctx, client, teardown := setup(t, &asyncWithInfoTimeoutHandler{})
+	defer teardown()
+
+	type testcase struct {
+		name         string
+		timeout      time.Duration
+		setOnHeader  bool
+		setOnContext bool
+		validator    func(t *testing.T, handle *OperationHandle[*LazyValue], info *OperationInfo, err error)
+	}
+	cases := []testcase{
+		{
+			name:         "time_out: set on context",
+			timeout:      1 * time.Millisecond,
+			setOnHeader:  false,
+			setOnContext: true,
+			validator: func(t *testing.T, handle *OperationHandle[*LazyValue], info *OperationInfo, err error) {
+				require.ErrorContains(t, err, "context deadline exceeded")
+			},
+		},
+		{
+			name:         "time_out: set on header",
+			timeout:      1 * time.Millisecond,
+			setOnHeader:  true,
+			setOnContext: false,
+			validator: func(t *testing.T, handle *OperationHandle[*LazyValue], info *OperationInfo, err error) {
+				require.ErrorContains(t, err, "handler exceeded request timeout of 1ms")
+			},
+		},
+		{
+			name:         "time_out: set on context and header",
+			timeout:      1 * time.Millisecond,
+			setOnHeader:  true,
+			setOnContext: true,
+			validator: func(t *testing.T, handle *OperationHandle[*LazyValue], info *OperationInfo, err error) {
+				require.ErrorContains(t, err, "context deadline exceeded")
+			},
+		},
```

Comment:
I would test that the header overwrites the context deadline and not go beyond that.

Reply:
You'd want contextTimeout and headerTimeout to be set separately to test this.

Reply:
You only want 3 tests for every API:

1. context deadline is propagated to the server
2. request header overwrites header from context deadline
3. no context deadline is set on server if there's no context deadline in client (with a special case for GetResult with `wait`).